pkg/services/ngalert/store/alert_rule.go

```diff
@@ -0,0 +1,443 @@
+package store
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/grafana/grafana/pkg/models"
+
+	"github.com/grafana/grafana/pkg/services/dashboards"
+
+	apimodels "github.com/grafana/alerting-api/pkg/api"
+	ngmodels "github.com/grafana/grafana/pkg/services/ngalert/models"
+	"github.com/grafana/grafana/pkg/services/sqlstore"
+	"github.com/grafana/grafana/pkg/util"
+)
+
+type UpdateRuleGroupCmd struct {
+	OrgID           int64
+	NamespaceUID    string
+	RuleGroup       string
+	RuleGroupConfig apimodels.PostableRuleGroupConfig
+}
+
+type UpsertRule struct {
+	Existing *ngmodels.AlertRule
+	New      ngmodels.AlertRule
+}
+
+// Store is the interface for persisting alert rules and instances
+type RuleStore interface {
+	DeleteAlertRuleByUID(*ngmodels.DeleteAlertRuleByUIDCommand) error
+	DeleteNamespaceAlertRules(*ngmodels.DeleteNamespaceAlertRulesCommand) error
+	DeleteRuleGroupAlertRules(*ngmodels.DeleteRuleGroupAlertRulesCommand) error
+	GetAlertRuleByUID(*ngmodels.GetAlertRuleByUIDQuery) error
+	GetAlertRules(query *ngmodels.ListAlertRulesQuery) error
+	GetOrgAlertRules(query *ngmodels.ListAlertRulesQuery) error
+	GetNamespaceAlertRules(query *ngmodels.ListNamespaceAlertRulesQuery) error
+	GetRuleGroupAlertRules(query *ngmodels.ListRuleGroupAlertRulesQuery) error
+	GetNamespaceUIDBySlug(string, int64, *models.SignedInUser) (string, error)
+	GetNamespaceByUID(string, int64, *models.SignedInUser) (string, error)
+	UpsertAlertRules([]UpsertRule) error
+	UpdateRuleGroup(UpdateRuleGroupCmd) error
+	GetAlertInstance(*ngmodels.GetAlertInstanceQuery) error
+	ListAlertInstances(cmd *ngmodels.ListAlertInstancesQuery) error
+	SaveAlertInstance(cmd *ngmodels.SaveAlertInstanceCommand) error
+	ValidateAlertRule(ngmodels.AlertRule, bool) error
+}
+
+func getAlertRuleByUID(sess *sqlstore.DBSession, alertRuleUID string, orgID int64) (*ngmodels.AlertRule, error) {
+	// we consider optionally enabling some caching
+	alertRule := ngmodels.AlertRule{OrgID: orgID, UID: alertRuleUID}
+	has, err := sess.Get(&alertRule)
+	if !has {
+		return nil, ngmodels.ErrAlertRuleNotFound
+	}
+	if err != nil {
+		return nil, err
+	}
```

Comment:
The error check should come first, otherwise there be dragons.
```suggestion
	if err != nil {
		return nil, err
	}
	if !has {
		return nil, ngmodels.ErrAlertRuleNotFound
	}
```